sync2/rangesync/p2p_test.go

```diff
@@ -0,0 +1,399 @@
+	}
+}
+
+func p2pRequesterGetter(t *testing.T) getRequesterFunc {
+	mesh, err := mocknet.FullMeshConnected(2)
+	require.NoError(t, err)
+	proto := "itest"
+	opts := []server.Opt{
+		server.WithRequestSizeLimit(100_000_000),
+		server.WithTimeout(10 * time.Second),
+		server.WithLog(zaptest.NewLogger(t)),
+	}
+	return func(
+		name string,
+		handler server.StreamHandler,
+		peers ...rangesync.Requester,
+	) (rangesync.Requester, p2p.Peer) {
+		if len(peers) == 0 {
+			return server.New(mesh.Hosts()[0], proto, handler, opts...), mesh.Hosts()[0].ID()
+		}
+		s := server.New(mesh.Hosts()[1], proto, handler, opts...)
+		require.Eventually(t, func() bool {
+			for _, h := range mesh.Hosts()[0:] {
+				if len(h.Mux().Protocols()) == 0 {
+					return false
+				}
+			}
+			return true
+		}, time.Second, 10*time.Millisecond)
+		return s, mesh.Hosts()[1].ID()
+	}
+}
+
+type syncTracer struct {
+	dumb          bool
+	receivedItems int
+	sentItems     int
+}
+
+var _ rangesync.Tracer = &syncTracer{}
+
+func (tr *syncTracer) OnDumbSync() {
+	tr.dumb = true
+}
+
+func (tr *syncTracer) OnRecent(receivedItems, sentItems int) {
+	tr.receivedItems += receivedItems
+	tr.sentItems += sentItems
+}
+
+// fakeRecentSet is a wrapper around OrderedSet that keeps track of the time when each
+// item was added to the set according to the specified clock.
+// It is used to test recent sync.
+type fakeRecentSet struct {
+	rangesync.OrderedSet
+	timestamps map[string]time.Time
+	clock      clockwork.Clock
+}
+
+var _ rangesync.OrderedSet = &fakeRecentSet{}
+
+var startDate = time.Date(2024, 8, 29, 18, 0, 0, 0, time.UTC)
+
+// registerAll assigns timestamps to all the items currently in the set.
+func (frs *fakeRecentSet) registerAll(_ context.Context) error {
+	frs.timestamps = make(map[string]time.Time)
+	t := startDate
+	items, err := rangesync.CollectSetItems(frs.OrderedSet)
+	if err != nil {
+		return err
+	}
+	for _, v := range items {
+		frs.timestamps[string(v)] = t
+		t = t.Add(time.Second)
+	}
+	return nil
+}
+
+// Receive implements OrderedSet.
+func (frs *fakeRecentSet) Receive(k rangesync.KeyBytes) error {
+	if err := frs.OrderedSet.Receive(k); err != nil {
+		return err
+	}
+	frs.timestamps[string(k)] = frs.clock.Now()
+	return nil
+}
+
+// Recent implements OrderedSet.
+func (frs *fakeRecentSet) Recent(since time.Time) (rangesync.SeqResult, int) {
+	var items []rangesync.KeyBytes
+	items, err := rangesync.CollectSetItems(frs.OrderedSet)
+	if err != nil {
+		return rangesync.ErrorSeqResult(err), 0
+	}
+	for _, k := range items {
+		if !frs.timestamps[string(k)].Before(since) {
+			items = append(items, k)
+		}
+	}
+	return rangesync.SeqResult{
+		Seq: func(yield func(rangesync.KeyBytes) bool) {
+			for _, h := range items {
+				if !yield(h) {
+					return
+				}
+			}
+		},
+		Error: rangesync.NoSeqError,
+	}, len(items)
+}
+
+func testWireSync(t *testing.T, getRequester getRequesterFunc) {
+	for _, tc := range []struct {
```

Comment:
`t.Helper()` makes most sense for real test helpers like `verifyThisSmallThing`, so that when something inside that `verifyThisSmallThing` fails the uninteresting stack frames from inside `verifyThisSmallThing` are skipped and we see the line number of the place from which `verifyThisSmallThing` is called.

Here we have a different situation, and we're very much interested in what happens inside `testWireSync`, so the effect of hiding the related stack frames is undesirable.

In this case, test body that must be used with different `getRequester` funcs was extracted into `testWireSync` func to avoid too much of rightward drift as compared to another outer `for _, tc := range []struct` or smth like that (could also have testWireSync as a nested func but not sure if this really has any benefits).
Repeating every test case twice in the code here I also can't agree with b/c the important part is to make sure exactly same test cases pass for the fake requester and the real one. If only real requester fails to pass the test, we have a bug in the requester impl. If both requesters fail, the bug may be in the set reconciliation mechanism (`RangeSetReconciler`)